### ISSUE TITLE
Inconsistent casting flexibility between Query and QueryAsync

### ISSUE THREAD

**Issue by ejball** [StackExchange/dapper-dot-net#550]
`Query<long>` and `QueryAsync<long>` have inconsistent behavior in at least one instance with MySQL. The former will implicitly cast `ulong` to `long`, but the latter throws `InvalidCastException`. I don't know which is correct, but the inconsistency is unfortunate, e.g. when porting from sync to async.

```csharp
async Task Main()
{
    using (var connection = new MySqlConnection(connectionString))
    {
        connection.Open();
        connection.Execute(@"
            DROP TEMPORARY TABLE IF EXISTS xyzzy;
            CREATE TEMPORARY TABLE xyzzy (
                id BIGINT NOT NULL AUTO_INCREMENT,
                value BIGINT NOT NULL,
                    PRIMARY KEY (id)
                ) ENGINE=InnoDB;");

        string sql = @"INSERT INTO xyzzy (value) VALUES (42); SELECT LAST_INSERT_ID();";

        // both succeed as expected
        Console.WriteLine(connection.Query<ulong>(sql).Single());
        Console.WriteLine((await connection.QueryAsync<ulong>(sql)).Single());

        // doesn't throw InvalidCastException
        Console.WriteLine(connection.Query<long>(sql).Single());

        // throws InvalidCastException
        Console.WriteLine((await connection.QueryAsync<long>(sql)).Single());
    }
}
```

**Comment by ejball** [comment]
We've seen a similar problem with a `BIT` value -- `Query<bool>` works fine but `QueryAsync<bool>` fails to cast the `ulong`.

**Comment by mgravell** [comment]
Which version are you using? I know there have been changes in this area
over the "alpha" builds; if you aren't already, I recommend trying against
the now fully-released 1.50.* builds. If it doesn't work there, then
definitely a bug and needs fixing. So again: which version are you using?

On 11 July 2016 at 23:22, Ed Ball <notifications@github.com> wrote:

> We've seen a similar problem with a BIT value -- Query<bool> works fine
> but QueryAsync<bool> fails to cast the ulong.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/StackExchange/dapper-dot-net/issues/550#issuecomment-231883762>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AABDsNVxrcAIXyYS3Hmhm-FutA8PtgOyks5qUsIpgaJpZM4JDTaS>
> .
>



-- 
Regards,

Marc

**Comment by ejball** [comment]
Thanks, Marc! I should have thought to check that. I was still on 1.42; the latest release works as expected.
